cpukit/score/src/schedulerstrongapa.c

```diff
@@ -3,339 +3,820 @@
+    node = (Scheduler_strong_APA_Node *) STRONG_SCHEDULER_NODE_OF_CHAIN( next );
+
+    if (
+      _Scheduler_SMP_Node_state( &node->Base.Base ) ==
+      SCHEDULER_SMP_NODE_READY
+    ) {
+      return true;
+    }
+
+    next = _Chain_Next( next );
+  }
+
+  return false;
 }
 
-static void _Scheduler_strong_APA_Move_from_ready_to_scheduled(
+static inline void _Scheduler_strong_APA_Set_scheduled(
+  Scheduler_strong_APA_Context *self,
+  Scheduler_Node                *executing,
+  const Per_CPU_Control         *cpu
+)
+{
+  self->CPU[ _Per_CPU_Get_index( cpu ) ].executing = executing;
+}
+
+static inline Scheduler_Node *_Scheduler_strong_APA_Get_scheduled(
+  const Scheduler_strong_APA_Context *self,
+  const Per_CPU_Control               *cpu
+)
+{
+  return self->CPU[ _Per_CPU_Get_index( cpu ) ].executing;
+}
+
+static inline void _Scheduler_strong_APA_Allocate_processor(
   Scheduler_Context *context,
-  Scheduler_Node    *ready_to_scheduled
+  Scheduler_Node    *scheduled_base,
+  Scheduler_Node    *victim_base,
+  Per_CPU_Control   *victim_cpu
 )
 {
+  Scheduler_strong_APA_Node    *scheduled;
   Scheduler_strong_APA_Context *self;
-  Scheduler_strong_APA_Node    *node;
-  Priority_Control              insert_priority;
 
+  (void) victim_base;
+
+  scheduled = _Scheduler_strong_APA_Node_downcast( scheduled_base );
   self = _Scheduler_strong_APA_Get_self( context );
-  node = _Scheduler_strong_APA_Node_downcast( ready_to_scheduled );
+  
+  _Scheduler_strong_APA_Set_scheduled( self, scheduled_base, victim_cpu );
 
-  _Scheduler_priority_Ready_queue_extract(
-    &node->Base.Base.Node.Chain,
-    &node->Ready_queue,
-    &self->Bit_map
+  _Scheduler_SMP_Allocate_processor_exact(
+    context,
+    &( scheduled->Base.Base ),
+    NULL,
+    victim_cpu
   );
-  insert_priority = _Scheduler_SMP_Node_priority( &node->Base.Base );
+}
+
+/*
+ * Finds and returns the highest ready node present by accessing the
+ * _Strong_APA_Context->CPU with front and rear values.
+ */
+static inline Scheduler_Node * _Scheduler_strong_APA_Find_highest_ready(
+  Scheduler_strong_APA_Context *self,
+  uint32_t                      front,
+  uint32_t                      rear
+)
+{
+  Scheduler_Node              *highest_ready;
+  Scheduler_strong_APA_CPU    *CPU;
+  const Chain_Node            *tail;
+  Chain_Node                  *next;
+  uint32_t                     index_assigned_cpu;
+  uint32_t                     index_curr_cpu;
+  Scheduler_strong_APA_Node   *node;
+  Priority_Control             min_priority_num;
+  Priority_Control             curr_priority;
+  Per_CPU_Control             *assigned_cpu;
+  Scheduler_SMP_Node_state     curr_state;
+  Per_CPU_Control             *curr_CPU;
+  bool                         first_task;
+
+  CPU = self->CPU;
+   /*
+    * When the first task accessed has nothing to compare its priority against
+    * So, it is the task with the highest priority witnessed so far.
+   */
+  first_task = true;
```

Comment:
Assigned min_priority_num = -1.

Reply:
although this works, it is not great to rely on automatic casting of a signed integer to the unsigned max. There is PRIORITY_DEFAULT_MAXIMUM but that is not valid in some SMP Schedulers, because the priority values can range higher than 255. You could maybe use UINT64_MAX

Reply:
Assigned min_priority_num = UINT64_MAX. I have not been compiling and making sure my changes work right now, I will do that later before making big changes (in the design or logic - like breaking down functions). Hopefully, these minor fixes should not break anything.

Reply:
Assigned UINT64_MAX and checked that it works.